tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java

```diff
@@ -332,7 +342,10 @@ public void updateStrategy() {
 		}
 
 		if (updated) {
-			strategy = FederationEvaluationStrategyFactory.getEvaluationStrategy(type, federationContext);
+			// TODO problem here: we need to (optionally) inject a Dataset object into the EvaluationStrategy, which is
+			// query-specific and determines what parts of the store are relevant for the query. However this part of
+			// the code assumes that we can instantiate an EvaluationStrategy once and then reuse it for each query.
```

Comment:
@aschwarte10 I could use your eyes on this problem. FedX is currently designed to instantiate the EvaluationStrategy once (unless the config changes). But in RDF4J Sails, it is generally assumed that each query gets its own EvaluationStategy - that is why the `Dataset` object is passed in on the constructor. I believe that this is caused some of the named graph queries to fail on FedX. I'd like to hear your opinion on the best way forward: can we just remove the assumption that the strategy is defined once and let FedX instantiate a new strategy for each query? 

I'm a little way of doing something like just setting/overwriting the Dataset  assigned to the strategy on each query, as I'm not sure what the consequences would be in a multithreaded environment.

Reply:
Hmm, I see. I wasn't aware of this assumption when initially implementing FedX.

Generally, I do not see a bigger problem in instantiating the strategy for each query (except for the amount of objects). However, quickly scanning the code and the references to where the strategy is passed around, such a change would be a huge refactoring.

I am not sure if at this time right before the release it is such a good idea to do this refactoring. If at all, I would suggest to do it outside of this PR to have a cleaner and easier change.

However, I think we can cope without it: the query specific QueryInfo object always has a reference to the externally passed Dataset. And this QueryInfo is available basically everywhere at evaluation time. This also how all the other named graph queries are working.

Let's synch offline on this and come to the best conclusion.

Note: I will probably re-order your last commit on top (i.e. a force-push), to have some other must changes integrated. The we have the possibility to decide on this special thing

Reply:
I agree a full refactoring is out of scope (and fwiw, the assumption I'm talking about isn't particularly well documented or obvious, so not blaming your design here). I may have an idea for a quick workaround fix. If I can get that to work I'll push it, otherwise I'll revert the WIP changes I made and log this as a known bug.